spark/src/main/scala/com/gu/tableversions/spark/VersionedDataset.scala

```diff
@@ -117,24 +118,21 @@ object VersionedDataset {
   /**
     * Write the given partitioned dataset, storing each partition in the associated path.
     */
-  private[spark] def writeVersionedPartitions[T](dataset: Dataset[T], partitionPaths: Map[Partition, URI]): Unit = {
-    // This is a slow and inefficient implementation that writes each partition in sequence,
-    // we can look into a more performant solution later.
-
-    def filteredForPartition(partition: Partition): Dataset[T] =
-      partition.columnValues.foldLeft(dataset) {
-        case (filteredDataset, partitionColumn) =>
-          filteredDataset.where(s"${partitionColumn.column.name} = '${partitionColumn.value}'")
-      }
-
-    val datasetsWithPaths: Map[Dataset[T], URI] =
-      partitionPaths.map { case (partition, path) => filteredForPartition(partition) -> path }
+  private[spark] def writeVersionedPartitions[T](
+      dataset: Dataset[T],
+      table: TableDefinition,
+      partitionPaths: Map[Partition, URI])(implicit spark: SparkSession): Unit = {
 
-    datasetsWithPaths.foreach {
-      case (datasetForPartition, partitionPath) =>
-        datasetForPartition.write
-          .parquet(partitionPath.toString) // TODO: Take in format parameter. Or a dataset writer?
+    VersionPaths.flattenMap(partitionPaths.keys.toList) foreach {
+      case (k, v) => spark.sparkContext.hadoopConfiguration.set(k, v)
     }
+
+    val partitions = table.partitionSchema.columns.map(_.name)
+
+    dataset.write
+      .mode(SaveMode.Append)
+      .partitionBy(partitions: _*)
+      .parquet(VersionedFileSystem.SCHEME + "://" + table.location.getPath)
```

Comment:
Probably worth leaving in the comment about that points out that the format needs to be customisable?

Reply:
ouch, indeed!